core/src/banking_trace.rs

```diff
@@ -0,0 +1,510 @@
+use {
+    crate::sigverify::SigverifyTracerPacketStats,
+    bincode::serialize_into,
+    chrono::{DateTime, Local},
+    crossbeam_channel::{unbounded, Receiver, SendError, Sender, TryRecvError},
+    rolling_file::{RollingCondition, RollingConditionBasic, RollingFileAppender},
+    solana_perf::{
+        packet::{to_packet_batches, PacketBatch},
+        test_tx::test_tx,
+    },
+    solana_sdk::slot_history::Slot,
+    std::{
+        fs::{create_dir_all, remove_dir_all},
+        io::{self, Write},
+        path::PathBuf,
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc,
+        },
+        thread::{self, sleep, JoinHandle},
+        time::{Duration, SystemTime},
+    },
+    tempfile::TempDir,
+    thiserror::Error,
+};
+
+pub type BankingPacketBatch = Arc<(Vec<PacketBatch>, Option<SigverifyTracerPacketStats>)>;
+pub type BankingPacketSender = TracedSender;
+pub type BankingPacketReceiver = Receiver<BankingPacketBatch>;
+pub type TracerThreadResult = Result<(), TraceError>;
+
+#[derive(Error, Debug)]
+pub enum TraceError {
+    #[error("IO Error: {0}")]
+    IoError(#[from] std::io::Error),
+
+    #[error("Serialization Error: {0}")]
+    SerializeError(#[from] bincode::Error),
+
+    #[error("Integer Cast Error: {0}")]
+    IntegerCastError(#[from] std::num::TryFromIntError),
+
+    #[error("Trace size is too small (must be larger than {1}): {0}")]
+    TooSmallTraceSize(u64, u64),
+}
+
+const BASENAME: &str = "events";
+const TRACE_FILE_ROTATE_COUNT: u64 = 14; // target 2 weeks retention under normal load
+const TRACE_FILE_WRITE_INTERVAL_MS: u64 = 100;
+const BUF_WRITER_CAPACITY: usize = 10 * 1024 * 1024;
+pub const TRACE_FILE_DEFAULT_ROTATE_BYTE_THRESHOLD: u64 = 1024 * 1024 * 1024;
+pub const EMPTY_BANKING_TRACE_SIZE: u64 = 0;
+pub const DEFAULT_BANKING_TRACE_SIZE: u64 =
+    TRACE_FILE_DEFAULT_ROTATE_BYTE_THRESHOLD * TRACE_FILE_ROTATE_COUNT;
+
+#[allow(clippy::type_complexity)]
+#[derive(Debug)]
+pub struct BankingTracer {
+    enabled_tracer: Option<(
+        Sender<TimedTracedEvent>,
+        Option<JoinHandle<TracerThreadResult>>,
+        Arc<AtomicBool>,
+    )>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TimedTracedEvent(std::time::SystemTime, TracedEvent);
+
+#[derive(Serialize, Deserialize, Debug)]
+enum TracedEvent {
+    Bank(Slot, u32, BankStatus, usize),
```

Comment:
> minor: consider including Leader in the name of this variant - we only get these events at start/end of our leader slot/bank(s).

imo, i generally prefer shorter name if the surrounding context should suffice. also, naming consistency is important. so, if we would follow that prefixing logic, i'd rename the existing `BankStart` to `LeaderBankStart`. etc. again, i think it's a bit too wordy as its use context should clarity it's related to banking/leader.

> Wondering if we need/want the bank creation time for bank start events? That'd give us information about when it actually "started" vs when the thread(s) detected it started.

It's not strictly needed. Now that I had to introduce `BlockAndBankHash`which is more akin to _bank creation time_, instead of _bank detection time_. Combined with imprecise semantics of `LeaderSlotMetricsTracker`, I just decided to remove it for now: 567e2a2327f68f28e5a81c231c28dff40a95e350
 
> Haven't seen your replay tool, so not sure if that information is valuable for the usecase.

fyi, here's the draft simulation code with good (hopefully!) docs: https://github.com/ryoqun/solana/pull/2/files#r1070495835